### PR TITLE
[embedded] Avoid a crash on location-less SIL functions in PerformanceDiagnostics

### PR DESCRIPTION
**Pull request by kubamracek**
Turns out that SILGen can generate calls to things like `swift_readAtKeyPath`, which end up being location-less SILFunctions. Let's avoid crashing on those in PerformanceDiagnostics. The attached test case previously failed with:
```
Assertion failed: (DebugScope && "no scope/location"), function getLocation, file SILFunction.h, line 1086.
```

rdar://125724841

### CODE REVIEW

File: lib/SILOptimizer/Mandatory/PerformanceDiagnostics.cpp [hunk 1]
```diff
@@ -810,15 +810,18 @@ class PerformanceDiagnosticsPass : public SILModuleTransform {
...
-        auto func = function.getLocation().getAsASTNode<AbstractFunctionDecl>();
-        if (func) {
-          if (isa<DestructorDecl>(func) || isa<ConstructorDecl>(func)) {
-            constructorsAndDestructors.push_back(&function);
-            continue;
-          }
-          if (getMethodDispatch(func) == MethodDispatch::Class) {
-            vtableMembers.push_back(&function);
-            continue;
+        if (function.hasLocation()) {
```
Comment (eeckstein) [inline lib/SILOptimizer/Mandatory/PerformanceDiagnostics.cpp:813]: Can you add a short comment saying why there are functions with no locations and why it's okay to not check those?

**Comment by kubamracek** [comment]
@swift-ci please test

**Comment by kubamracek** [comment]
@swift-ci please test macOS platform

**Comment by kubamracek** [comment]
@swift-ci please test macOS platform

**Review by eeckstein — APPROVED** [review]
lgtm!

**Pull request merged**